tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc

```diff
@@ -0,0 +1,127 @@
+/* Copyright 2022 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#ifdef INTEL_MKL
+
+#include <initializer_list>
+
+#include "tensorflow/cc/ops/nn_ops.h"
+#include "tensorflow/cc/ops/nn_ops_internal.h"
+#include "tensorflow/core/common_runtime/kernel_benchmark_testlib.h"
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/graph/mkl_testlib.h"
+#include "tensorflow/core/graph/testlib.h"
+#include "tensorflow/core/public/session_options.h"
+
+namespace tensorflow {
+namespace {
+
+template <typename T>
+static void BM_Softmax(::testing::benchmark::State& state,
+                       std::initializer_list<int64_t> dims, int num_threads,
+                       const string& label, bool onednn) {
+  DataType dtype = DataTypeToEnum<T>::v();
+  TensorShape shape = TensorShape(dims);
+  Tensor input(dtype, shape);
+  input.flat<T>().setRandom();
+  Graph* g = new Graph(OpRegistry::Global());
+  if (onednn) {
+    test::graph::oneDNNSoftmax(g, test::graph::Constant(g, input));
+  } else {
+    auto root = Scope::NewRootScope().ExitOnError();
+    auto softmax = ops::Softmax(root, input);
+    TF_CHECK_OK(root.status());
+    TF_CHECK_OK(root.ToGraph(g));
+  }
+  SessionOptions opts;
+  opts.config.set_inter_op_parallelism_threads(1);
+  opts.config.set_intra_op_parallelism_threads(num_threads);
+  opts.config.set_use_per_session_threads(true);
+  opts.config.mutable_graph_options()
+      ->mutable_optimizer_options()
+      ->set_opt_level(OptimizerOptions::L0);
+  test::Benchmark("cpu", g, &opts, nullptr, nullptr, "",
+                  /*old_benchmark_api*/ false)
+      .Run(state);
+  state.SetItemsProcessed(shape.num_elements() * state.iterations());
+  state.SetLabel(label);
+}
+
+/*
+ * The following trick for number of variadic macro arguments is taken from:
+ * http://groups.google.com/group/comp.std.c/browse_thread/thread/77ee8c8f92e4a3fb/346fc464319b1ee5?pli=1
+ * The macro PP_NARG returns the number of arguments that have been passed to
+ * it.  Currrently, it is prepared to support upto 5 arguments. It is straight
+ * forward to extend to a higher number of arguments.
+ */
+#define PP_NARG(...) PP_NARG_(__VA_ARGS__, PP_RSEQ_N())
+#define PP_NARG_(...) PP_ARG_N(__VA_ARGS__)
+#define PP_ARG_N(_1, _2, _3, _4, _5, N, ...) N
+#define PP_RSEQ_N() 5, 4, 3, 2, 1, 0
```

Comment:
Unfortunately, I don't think we can take this code since it's not published under a license.

Can we explicitly specify the number of dims as a `BM_` macro parameter instead? E.g.,
```c++
#define BM_oneDNN_Softmax(dtype, num_threads, label, num_dims, ...)                    \
  static void JOIN(BM_oneDNN_Softmax_##dtype##_intraop_##num_threads##_dims, \
                   JOIN(CONCAT_DIMS, num_dims)(__VA_ARGS__))(    \
      ::testing::benchmark::State & state) {                                 \
    BM_Softmax<dtype>(state, {__VA_ARGS__}, num_threads, label, true);       \
  }  
  ...

// Example usage:
BM_Softmax(float, 1, "float32_ImageNet_batch_size_32", /*num_dims=*/2, 32, 1008);
```

(The code may not compile because I didn't test it.)

Reply:
Done.